File: src/analysis/categorizer.py
"""Job categorization for template library."""

import json
import logging
import os
import re
from typing import Dict, Any, Optional, List
from dataclasses import dataclass

import google.generativeai as genai

logger = logging.getLogger(__name__)


@dataclass
class JobCategoryResult:
    """Result of job categorization."""
    role_category: str
    extracted_sections: Dict[str, Any]
    quality_score: int  # 1-5

    def to_dict(self) -> Dict[str, Any]:
        return {
            "role_category": self.role_category,
            "extracted_sections": json.dumps(self.extracted_sections),
            "quality_score": self.quality_score,
        }


class JobCategorizer:
    """Categorize job postings and extract template sections."""

    # Common role categories in fintech
    ROLE_CATEGORIES = [
        "engineering-backend",
        "engineering-frontend",
        "engineering-fullstack",
        "engineering-mobile",
        "engineering-data",
        "engineering-ml",
        "engineering-devops",
        "engineering-security",
        "engineering-qa",
        "product-management",
        "product-design",
        "product-research",
        "marketing-growth",
        "marketing-product",
        "marketing-content",
        "marketing-brand",
        "sales",
        "customer-success",
        "customer-support",
        "operations",
        "finance",
        "legal-compliance",
        "hr-people",
        "data-analytics",
        "risk",
        "leadership",
        "other"
    ]

    CATEGORIZATION_PROMPT = """Analyze this job posting and categorize it for a job template library.

Job Title: {job_title}
Company: {company_name}
Description:
{description}

Provide analysis in JSON format:
{{
    "role_category": "<category from: {categories}>",
    "extracted_sections": {{
        "summary": "<1-2 sentence role summary>",
        "responsibilities": ["<responsibility 1>", "<responsibility 2>", ...],
        "requirements": ["<requirement 1>", "<requirement 2>", ...],
        "nice_to_have": ["<nice to have 1>", ...],
        "benefits": ["<benefit 1>", ...]
    }},
    "quality_score": <1-5 rating of how well-written this posting is>
}}

Quality scoring guide:
5 = Excellent: Clear, comprehensive, well-structured
4 = Good: Most sections present, clear language
3 = Average: Basic information, some gaps
2 = Below average: Vague or missing key sections
1 = Poor: Minimal useful information

Respond ONLY with valid JSON."""

    def __init__(self, api_key: Optional[str] = None, model: str = "gemini-2.0-flash"):
        """Initialize the categorizer with Gemini API."""
        self.api_key = api_key or os.getenv("GEMINI_API_KEY")
        if not self.api_key:
            raise ValueError("GEMINI_API_KEY not found in environment")

        genai.configure(api_key=self.api_key)
        self.model = genai.GenerativeModel(model)

    def categorize_posting(
        self,
        job_title: str,
        company_name: str,
        description: str
    ) -> Optional[JobCategoryResult]:
        """Categorize a job posting and extract template sections."""
        try:
            prompt = self.CATEGORIZATION_PROMPT.format(
                job_title=job_title,
                company_name=company_name,
                description=description[:5000] if description else "No description",
                categories=", ".join(self.ROLE_CATEGORIES)
            )

            response = self.model.generate_content(
                prompt,
                generation_config=genai.GenerationConfig(
                    temperature=0.2,
                    max_output_tokens=1500,
                )
            )

            response_text = response.text.strip()

            # Handle markdown code blocks
            if response_text.startswith("```"):
                lines = response_text.split("\n")
                response_text = "\n".join(lines[1:-1])

            result = json.loads(response_text)

            return JobCategoryResult(
                role_category=result.get("role_category", "other"),
                extracted_sections=result.get("extracted_sections", {}),
                quality_score=result.get("quality_score", 3)
            )

        except json.JSONDecodeError as e:
            logger.error(f"Failed to parse categorization response: {e}")
            return None
        except Exception as e:
            logger.error(f"Error categorizing posting '{job_title}': {e}")
            return None

    def quick_categorize(self, job_title: str) -> str:
        """Quick categorization based on job title alone (no API call)."""
        title_lower = job_title.lower()

        # Engineering categories
        if any(w in title_lower for w in ["backend", "server", "api"]):
            return "engineering-backend"
        if any(w in title_lower for w in ["frontend", "front-end", "ui ", "react", "vue", "angular"]):
            return "engineering-frontend"
        if any(w in title_lower for w in ["full stack", "fullstack", "full-stack"]):
            return "engineering-fullstack"
        if any(w in title_lower for w in ["mobile", "ios", "android", "swift", "kotlin"]):
            return "engineering-mobile"
        if any(w in title_lower for w in ["data engineer", "etl", "pipeline"]):
            return "engineering-data"
        if any(w in title_lower for w in ["machine learning", "ml ", "ai ", "deep learning"]):
            return "engineering-ml"
        if any(w in title_lower for w in ["devops", "sre", "infrastructure", "platform", "cloud"]):
            return "engineering-devops"
        if any(w in title_lower for w in ["security", "infosec", "cybersecurity"]):
            return "engineering-security"
        if any(w in title_lower for w in ["qa ", "quality", "test engineer", "sdet"]):
            return "engineering-qa"
        if any(w in title_lower for w in ["software", "engineer", "developer"]) and not any(w in title_lower for w in ["data", "ml", "ai"]):
            return "engineering-fullstack"

        # Product categories
        if any(w in title_lower for w in ["product manager", "product lead", "pm ", "product owner"]):
            return "product-management"
        if any(w in title_lower for w in ["product design", "ux ", "ui/ux", "user experience"]):
            return "product-design"
        if any(w in title_lower for w in ["user research", "ux research"]):
            return "product-research"

        # Marketing categories
        if any(w in title_lower for w in ["growth", "acquisition", "performance marketing"]):
            return "marketing-growth"
        if any(w in title_lower for w in ["product market", "pmm"]):
            return "marketing-product"
        if any(w in title_lower for w in ["content", "copywriter", "editorial"]):
            return "marketing-content"
        if any(w in title_lower for w in ["brand", "creative"]):
            return "marketing-brand"
        if "marketing" in title_lower:
            return "marketing-growth"

        # Other categories
        if any(w in title_lower for w in ["sales", "account executive", "business development", "bdm"]):
            return "sales"
        if any(w in title_lower for w in ["customer success", "csm"]):
            return "customer-success"
        if any(w in title_lower for w in ["support", "customer service"]):
            return "customer-support"
        if any(w in title_lower for w in ["operations", "ops "]):
            return "operations"
        if any(w in title_lower for w in ["finance", "accounting", "controller"]):
            return "finance"
        if any(w in title_lower for w in ["legal", "compliance", "regulatory"]):
            return "legal-compliance"
        if any(w in title_lower for w in ["hr ", "people", "recruiter", "talent"]):
            return "hr-people"
        if any(w in title_lower for w in ["analyst", "analytics", "data scientist", "bi "]):
            return "data-analytics"
        if any(w in title_lower for w in ["risk", "fraud"]):
            return "risk"
        if any(w in title_lower for w in ["vp ", "director", "head of", "chief", "cto", "cfo", "ceo"]):
            return "leadership"

        return "other"

    def get_templates_by_category(
        self,
        postings: List[Dict[str, Any]],
        category: str,
        top_n: int = 5
    ) -> List[Dict[str, Any]]:
        """Get the top N templates for a given category, sorted by quality score."""
        matching = [p for p in postings if p.get("role_category") == category]
        sorted_postings = sorted(
            matching,
            key=lambda x: x.get("quality_score", 0),
            reverse=True
        )
        return sorted_postings[:top_n]

File: web/lib/analysis/categorizer.ts
import { GoogleGenerativeAI } from "@google/generative-ai";
import { recordUsage, type OnUsage, type GeminiUsageMetadata } from "@/lib/ai/gemini-meter";
import { writeUsageEvent } from "@/lib/ai/gemini-telemetry";
import { resolveProvider } from "@/lib/ai/providers/registry";
import { openAiCompatGenerate } from "@/lib/ai/providers/openai-compat";
import { log } from "@/lib/log";

/**
 * Job Categorizer for Template Library
 * 
 * Logic ported from src/analysis/categorizer.py to TypeScript for Vercel deployment.
 * Uses Gemini Flash for AI-based categorization and template extraction.
 */

// Fintech-specific role categories (must match function-categories.ts)
// Updated to 45 categories across 7 groups for better fintech competitive intelligence
export const ROLE_CATEGORIES = [
  // Group 1: Engineering (10 categories)
  "engineering-backend",
  "engineering-frontend",
  "engineering-fullstack",
  "engineering-mobile",
  "engineering-data",
  "engineering-ai-ml",
  "engineering-platform-sre-devops",
  "engineering-security",
  "engineering-qa",
  "engineering-management",
  // Group 2: Product & Design (5 categories)
  "product-management",
  "product-design-ux",
  "product-research",
  "technical-program-management",
  "technical-writing",
  // Group 3: Data & Analytics (3 categories)
  "data-science",
  "data-analytics-bi",
  "analytics-engineering",
  // Group 4: Risk, Legal & Compliance (6 categories)
  "compliance",
  "aml-financial-crime",
  "fraud-trust-safety",
  "legal",
  "regulatory-affairs",
  "risk-management",
  // Group 5: Go-To-Market (8 categories)
  "sales-account-executives",
  "account-management-customer-success",
  "customer-support",
  "business-development-partnerships",
  "solutions-engineering",
  "revenue-operations",
  "marketing-growth-performance",
  "marketing-product-brand",
  "developer-relations",
  // Group 6: Finance & Strategy (5 categories)
  "finance-accounting",
  "strategic-finance-fpa",
  "capital-markets-treasury",
  "corporate-development-strategy",
  "investor-relations",
  // Group 7: Operations & People (8 categories)
  "customer-support-cx",
  "customer-operations",
  "people-ops-hr",
  "talent-acquisition-recruiting",
  "it-internal-systems",
  "business-operations",
  "executive-leadership",
  "administrative",
  // Other
  "other"
] as const;

export type RoleCategory = typeof ROLE_CATEGORIES[number];

export interface ExtractedSections {
  summary: string;
  responsibilities: string[];
  requirements: string[];
  nice_to_have: string[];
  benefits: string[];
}

export interface JobCategoryResult {
  role_category: RoleCategory;
  extracted_sections: ExtractedSections;
  quality_score: number; // 1-5
}

const CATEGORIZATION_PROMPT = `Analyze this fintech job posting and categorize it using the fintech-specific taxonomy.

Job Title: {job_title}
Company: {company_name}
Description:
{description}

**Available Categories (45 total, organized into 7 groups):**

**Group 1: Engineering** (10 categories)
- engineering-backend, engineering-frontend, engineering-fullstack, engineering-mobile
- engineering-data, engineering-ai-ml, engineering-platform-sre-devops
- engineering-security, engineering-qa, engineering-management

**Group 2: Product & Design** (5 categories)
- product-management, product-design-ux, product-research
- technical-program-management, technical-writing

**Group 3: Data & Analytics** (3 categories)
- data-science, data-analytics-bi, analytics-engineering

**Group 4: Risk, Legal & Compliance** (6 categories)
- compliance, aml-financial-crime, fraud-trust-safety
- legal, regulatory-affairs, risk-management

**Group 5: Go-To-Market** (8 categories)
- sales-account-executives, account-management-customer-success, customer-support
- business-development-partnerships, solutions-engineering, revenue-operations
- marketing-growth-performance, marketing-product-brand, developer-relations

**Group 6: Finance & Strategy** (5 categories)
- finance-accounting, strategic-finance-fpa, capital-markets-treasury
- corporate-development-strategy, investor-relations

**Group 7: Operations & People** (8 categories)
- customer-support-cx, customer-operations, people-ops-hr
- talent-acquisition-recruiting, it-internal-systems, business-operations
- executive-leadership, administrative

**Other:** other

**Edge Case Rules (CRITICAL for accuracy):**
- SQL/Data Pipelines/ETL → engineering-data (Data Engineering), NOT data-analytics-bi
- Excel/Financial Modeling/Accounting → finance-accounting, NOT data-analytics-bi
- Product Engineer → engineering-fullstack (if technical) OR product-management (if strategic/PM-focused)
- Marketing Engineer → engineering-frontend or engineering-fullstack (technical role), NOT marketing
- Platform Engineering/SRE → engineering-platform-sre-devops (distinct from generic DevOps in mature fintechs)
- Risk Modeling/Credit Scoring → data-science (predictive modeling), NOT risk-management
- Fraud Detection Operations → fraud-trust-safety, NOT risk-management
- Compliance Operations/KYC Review → compliance, NOT legal
- Legal Counsel/Contracts → legal, NOT compliance
- Data Science (ML models for risk/credit) → data-science, NOT engineering-ai-ml
- Analytics Engineering (dbt, data modeling) → analytics-engineering, NOT engineering-data
- Customer Support (tickets/calls) → customer-support-cx, NOT account-management-customer-success
- Customer Success (retention/upsell) → account-management-customer-success, NOT customer-support-cx

**Key Distinctions:**
- Platform/SRE/DevOps: Infrastructure, cloud, reliability (engineering-platform-sre-devops)
- Data Engineering: Pipelines, warehousing, ETL (engineering-data)
- Data Science: Predictive modeling, ML for business (data-science)
- Data Analytics/BI: Reporting, dashboards, analysis (data-analytics-bi)
- Analytics Engineering: dbt, data modeling, transformation (analytics-engineering)
- Compliance: KYC, KYB, regulatory operations (compliance)
- Legal: Corporate counsel, contracts, litigation (legal)
- Risk Management: Credit risk, market risk, liquidity (risk-management)
- Fraud/Trust & Safety: Fraud detection, trust operations (fraud-trust-safety)

Provide analysis in JSON format:
{
    "role_category": "<category from: {categories}>",
    "extracted_sections": {
        "summary": "<1-2 sentence role summary>",
        "responsibilities": ["<responsibility 1>", "<responsibility 2>", ...],
        "requirements": ["<requirement 1>", "<requirement 2>", ...],
        "nice_to_have": ["<nice to have 1>", ...],
        "benefits": ["<benefit 1>", ...]
    },
    "quality_score": <1-5 rating of how well-written this posting is>
}

Quality scoring guide:
5 = Excellent: Clear, comprehensive, well-structured
4 = Good: Most sections present, clear language
3 = Average: Basic information, some gaps
2 = Below average: Vague or missing key sections
1 = Poor: Minimal useful information

Respond ONLY with valid JSON.`;


export interface CategorizePostingOptions {
  /**
   * Model override. Defaults to Gemini Flash (production). An eval id
   * (e.g. glm-5.2, deepseek-v4-flash) routes through the OpenAI-compatible
   * provider for the bake-off. See lib/ai/providers/registry.ts.
   */
  model?: string;
  /** Optional observer for token/cost usage (used by the eval harness). */
  onUsage?: OnUsage;
}

const DEFAULT_CATEGORIZE_MODEL = "gemini-flash-latest";

/**
 * Categorize a job posting and extract template sections.
 *
 * Uses Gemini Flash by default. Writes a `gemini_usage_events` row like every
 * other AI call site (this previously had no telemetry — see web/lib/ai/CLAUDE.md
 * "Telemetry"). The eval seam routes an eval model id to Fireworks.
 */
export async function categorizePosting(
  jobTitle: string,
  companyName: string,
  description: string,
  options?: CategorizePostingOptions
): Promise<JobCategoryResult | null> {
  const requestedModel = options?.model ?? DEFAULT_CATEGORIZE_MODEL;
  const provider = resolveProvider(requestedModel);
  const key = process.env.GEMINI_API_KEY;
  if (provider === "gemini" && !key) {
    log.error("GEMINI_API_KEY not configured");
    return null;
  }

  const prompt = CATEGORIZATION_PROMPT
    .replace("{job_title}", jobTitle)
    .replace("{company_name}", companyName)
    .replace("{description}", description.slice(0, 8000) || "No description")
    .replace("{categories}", ROLE_CATEGORIES.join(", "));

  try {
    let text: string;
    let usageMetadata: GeminiUsageMetadata | null | undefined;
    let modelServed = requestedModel;
    const startMs = Date.now();

    if (provider === "gemini") {
      const genAI = new GoogleGenerativeAI(key as string);

      // Using Gemini Flash (via `gemini-flash-latest`) as requested
      const model = genAI.getGenerativeModel({
        model: requestedModel,
        generationConfig: {
          temperature: 0.2,
          maxOutputTokens: 64000,
          responseMimeType: "application/json",
        },
      });

      const result = await model.generateContent({
        contents: [{ role: "user", parts: [{ text: prompt }] }],
      });
      text = result.response.text()?.trim() ?? "{}";
      usageMetadata = result.response.usageMetadata;
    } else {
      // Open-model evaluation path (eval-only; never a production default).
      const r = await openAiCompatGenerate({
        model: requestedModel,
        prompt,
        jsonMode: true,
        temperature: 0.2,
        maxOutputTokens: 8192,
        timeoutMs: 60000,
      });
      text = r.text?.trim() ?? "{}";
      usageMetadata = r.usageMetadata;
      modelServed = r.modelServed;
    }

    const usage = recordUsage({
      callSite: "categorizePosting",
      modelRequested: requestedModel,
      modelServed,
      groundingEnabled: false,
      usageMetadata,
      latencyMs: Date.now() - startMs,
      status: "ok",
      extra: { jobTitle, provider },
    });
    writeUsageEvent(usage);
    options?.onUsage?.(usage);

    const parsed = JSON.parse(text) as Record<string, unknown>;

    return {
      role_category: (parsed.role_category as RoleCategory) || "other",
      extracted_sections: (parsed.extracted_sections as ExtractedSections) || {
        summary: "",
        responsibilities: [],
        requirements: [],
        nice_to_have: [],
        benefits: []
      },
      quality_score: typeof parsed.quality_score === 'number' ? parsed.quality_score : 3
    };
  } catch (error) {
    log.error({ err: error }, `Error categorizing posting '${jobTitle}':`);
    return null;
  }
}
